chainforge/react-server/src/backend/ai.ts

```diff
@@ -221,6 +301,83 @@ export async function autofill(
   return new_items.slice(0, n);
 }
 
+/**
+ * Uses an LLM to interpret the pattern from the given table (columns and rows) and generate new rows following the pattern.
+ * @param input Object containing the columns and rows of the input table.
+ * @param n Number of new rows to generate.
+ * @param provider The LLM provider to use.
+ * @param apiKeys API keys required for the LLM query.
+ * @returns A promise resolving to an object containing updated columns and rows.
+ */
+export async function autofillTable(
+  input: { cols: string[]; rows: Row[] },
+  n: number,
+  provider: string,
+  apiKeys: Dict,
+): Promise<{ cols: string[]; rows: Row[] }> {
+
+  // Hash the arguments to get a unique id
+  const id = JSON.stringify([input, n]);
+
+  // Encode the input table to a markdown table
+  const encoded = encodeTable(input.cols, input.rows);
```

Comment:
In encodeTable, you will need to set an upward bound to how many rows/characters it will add. I'd say 20-30 rows is enough to see a pattern, but the better solution would *also* count the number of characters in each row and threshold on that smartly in case there's a lot of text in each row. Finally, the ideal solution would also sample those 20-30 rows at random---just in case the table is ordered in some way, and it's not possible to get a full taste of the large dataset. 

So you might have a lower bound ---e.g., it will always send 5 rows, regardless of the characters. Then, you could characters and keep counting until you either reach the upper bound threshold, say 30 rows, OR you hit the character limit (whatever comes first).